### ISSUE TITLE
Anchor labeling implementation question

### ISSUE THREAD

**Issue by KittyLiou** [facebookresearch/Detectron#753]
Hi, after I read Faster R-CNN paper I tried to understand how exactly the anchors are labeled by reading the code in `detectron/roi_data/rpn.py`
From the code below, my understanding is that if an anchor is labeled as foreground, its corresponding bounding box target will be the ground-truth that it has max IoU with. Am I right so far?
```
bbox_targets[fg_inds, :] = data_utils.compute_targets(
anchors[fg_inds, :], gt_boxes[anchor_to_gt_argmax[fg_inds], :]
)
```
But according to the paper, there are two cases that would make an anchor labeled as 'foreground':
1) the anchor with the highest Intersection-overUnion (IoU) overlap with a ground-truth box
2) an anchor that has an IoU overlap higher than 0.7 with any ground-truth box

My question is that what if an anchor is labeled as foreground due to rule 1 but it actually has its own max IoU with another ground-truth box?
For example, say there are only 2 anchors and 3 ground-truth boxes and `anchor_by_gt_overlap` looks like the below (row represents the anchors, column represents the ground-truth bounding boxes)
```
[[0.1, 0.4, 0.5], 
 [0.2, 0.1, 0.9]]
```
For anchor 0, it should be labeled as foreground according to rule 1 as it has max IoU with ground-truth 1
However, when computing the target of anchor 0, ground-truth 2 will be used instead of ground-truth 1 since for anchor 0 it actually has max IoU with ground-truth 2 rather than ground-truth 1.
This is where I'm confused.
Is this a bug or did I misunderstand some parts?

**Comment by xieshuqin** [comment]
I guess this should be fine. The binary label here just means objectiveness. So if an anchor is labeled with 1, it does contain some objects, regardless of the number of objects. But the box regression target is another story, it should find the most desirable target, which is the one with maximum overlap. I think the key here is that the binary label here is just an objectiveness term so it should be fine if multiple ground truths are matched.

**Comment by ppwwyyxx** [comment]
@KittyLiou I agree with your concerns and the it does not feel right when an anchor is labeled as positive because of ground-truth A, but has ground-truth B as its regression target.

However please also note that the labeling strategies are just heuristics, including what to label as positives,  what are the thresholds, etc. There isn't a definite answer on what is right or wrong. And the current strategy exists mainly because it works.

Your suggestion does make sense and perhaps it may improve performance in certain cases. However at least on COCO, I didn't see a performance difference from a quick experiment. I'll be interested to know if you find situations where it helps.
